
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/proto/dataplane/sai";

enum RpfGroupAttr {
  RPF_GROUP_ATTR_UNSPECIFIED = 0;
  RPF_GROUP_ATTR_RPF_INTERFACE_COUNT = 1;
  RPF_GROUP_ATTR_RPF_MEMBER_LIST = 2;
}

enum RpfGroupMemberAttr {
  RPF_GROUP_MEMBER_ATTR_UNSPECIFIED = 0;
  RPF_GROUP_MEMBER_ATTR_RPF_GROUP_ID = 1;
  RPF_GROUP_MEMBER_ATTR_RPF_INTERFACE_ID = 2;
}

message CreateRpfGroupRequest {
  uint64 switch = 1;
}

message CreateRpfGroupResponse {
  uint64 oid = 1;
}

message RemoveRpfGroupRequest {
  uint64 oid = 1;
}

message RemoveRpfGroupResponse {}

message GetRpfGroupAttributeRequest {
  uint64 oid = 1;
  RpfGroupAttr attr_type = 2;
}

message GetRpfGroupAttributeResponse {
  RpfGroupAttribute attr = 1;
}

message CreateRpfGroupMemberRequest {
  uint64 switch = 1;

  uint64 rpf_group_id = 2;
  uint64 rpf_interface_id = 3;
}

message CreateRpfGroupMemberResponse {
  uint64 oid = 1;
}

message RemoveRpfGroupMemberRequest {
  uint64 oid = 1;
}

message RemoveRpfGroupMemberResponse {}

message GetRpfGroupMemberAttributeRequest {
  uint64 oid = 1;
  RpfGroupMemberAttr attr_type = 2;
}

message GetRpfGroupMemberAttributeResponse {
  RpfGroupMemberAttribute attr = 1;
}

service RpfGroup {
  rpc CreateRpfGroup(CreateRpfGroupRequest) returns (CreateRpfGroupResponse) {}
  rpc RemoveRpfGroup(RemoveRpfGroupRequest) returns (RemoveRpfGroupResponse) {}
  rpc GetRpfGroupAttribute(GetRpfGroupAttributeRequest)
      returns (GetRpfGroupAttributeResponse) {}
  rpc CreateRpfGroupMember(CreateRpfGroupMemberRequest)
      returns (CreateRpfGroupMemberResponse) {}
  rpc RemoveRpfGroupMember(RemoveRpfGroupMemberRequest)
      returns (RemoveRpfGroupMemberResponse) {}
  rpc GetRpfGroupMemberAttribute(GetRpfGroupMemberAttributeRequest)
      returns (GetRpfGroupMemberAttributeResponse) {}
}
